### ISSUE TITLE
Tracking issue for normalizing prototype objects properly in CheckAccessControls

### ISSUE THREAD

**Issue by aravind-pg** [google/closure-compiler#1976]
(Note: this issue is purely for tracking/documentation purposes. I already have a fix locally.)

Suppose we have SubFoo extending Foo, so that SubFoo.prototype is an instance of Foo. We make a distinction between the particular instance of Foo that is SubFoo.prototype and a standard, canonical instance of Foo. We are currently confounding the two in the NTI case in `CheckAccessControls::normalizeClassType`, leading to spurious warnings in at least one place already.

Consider the following test case (eliding goog.provides and requires for clarity).
- `a.js`: 

``` javascript
/** @constructor */
var A = function() {}
/** @protected */
A.prototype.method = function() {}
```
- `b1.js`:

``` javascript
/** @constructor @extends {A} */
var B1 = function() {}
/** @override */
B1.prototype.method = function() {}
```
- `b2.js`:

``` javascript
/** @constructor @extends {A} */
var B2 = function() {}
/** @override */
B2.prototype.method = function() {}
```
- `c.js`:

``` javascript
/**
 * @param {!B1} b1
 * @constructor
 * @extends {B2}
 */
var C = function(b1) {
  var x = b1.method();
}
```

When run with NTI, this yields a spurious "Access to protected property `method` of `B1` not allowed here" CheckAccessControls warning.

Now what follows in the comments is a very concrete description of why this is happening and how my fix addresses it.

**Comment by aravind-pg** [comment]
What is happening:
- When analyzing the getprop `b1.method()`, we look up the superclass of `B1` that most recently overrode `method` -- or rather, we look for that superclass's prototype. This happens in `AccessControlUtils::getObjectType`.
- Specifically, we do this by ascending the prototype chain of `B1` until we find the prototype that first defines `method`. So we end up getting, in this case, `B1.prototype`.
- `B1.prototype` is a very specific instance of `A`. Specifically, in NTI it has type `A{method:function(this:B1):?}`, whereas the canonical instance of `A` is just `A`.
- To check if the protected property access is OK, we normalize the owner type, i.e. `B1.prototype` to its corresponding instance type, i.e. `A` [*, see caveat], and then check if the current class from which we are performing this property access, which is `C`, is a subtype of `A`.
- What was happening was that in the NTI case this normalization was not actually occurring. So instead of checking for `C <: A`, we were checking `C <: A{method:function(this:B1):?}`. But this is actually false, since in `C` `method` has type `function(this:B2):?` and not `function(this:B1):?` (with `B1, B2` having no subtype relationship).
  - In the OTI case, the `isPrototypeObject` codepath in `CheckAccessControls::normalizeClassType` does the normalization.
  - In the NTI case, `isPrototypeObject` is a stub that always returns false, so the normalization wasn't happening.

My fix addresses this by adding a `normalizeMaybePrototypeObject` API to ObjectTypeI that does this normalization more properly in both NTI and OTI. The NTI implementation is essentially just `getConstructor().getInstance()`, which does the trick.

------

(*) Caveat:
- Unfortunately the notion of "corresponding instance type" is not quite what I said above. For `SubFoo.prototype` it is `SubFoo` itself, not `Foo`.
- This is not actually as big of a deal as it may seem for subtyping purposes.
- The contract of `normalizeMaybePrototypeObject` is precisely to return this "corresponding instance type".
  - In OTI we are able to do this successfully using `getOwnerFunction`, which for `SubFoo.prototype` returns the constructor for `SubFoo`. So the implementation is just `getOwnerFunction().getInstance()`, and we meet the contract.
  - However in NTI we do not have a `getOwnerFunction` equivalent, since we do not store the "constructor" property on prototype objects. (This is a known issue, intended to be fixed.) So we have to make do by using `getConstructor` instead, which for `SubFoo.prototype` returns the constructor of `Foo` (not `SubFoo`). So the implementation is `getConstructor().getInstance()` which actually fails to meet the contract, returning `Foo` instead of `SubFoo` for `SubFoo.prototype`.
- From a subtyping point of view this means we are slightly more liberal in NTI. (Since `Bar <: SubFoo` => `Bar <: Foo`.)
- All things considered, this is not a huge caveat, since test cases that are affected by this fine difference are not too common in practice. The current fix is certainly an improvement on the state of affairs, and this caveat can be addressed later once we implement the "constructor" property.

**Comment by aravind-pg** [comment]
I wanted to note another wrinkle that makes this business still more complicated: because of the way `getOwnProp` works in NTI (different from its behavior in OTI), the caveat noted above actually becomes _necessary_ for the two to have the same behavior.

Specifically, `getOwnProp` in NTI considers properties on the prototype (of an object type) to be own props, whereas OTI does not. [Note: when I say `getOwnProp` I am referring to `getOwnPropertyDefSite` and the like in ObjectTypeI.]

Now, `getOwnProp` is used by `AccessControlUtils::getObjectType`, which (essentially) ascends the prototype chain until it finds the first type that defines the required property as an own prop, and returns that type. Because NTI considers prototype properties to be own props as well, it finds the first type earlier than OTI does.

To be concrete, consider the inheritance chain of `C -> B2 -> A` in the example. Suppose we do `getObjectType(C, "method")`, where `C` refers to the `C` instance type. Then this is what happens in OTI:
  1. Is "method" an own prop of C? No.
  2. Ascend the prototype chain, look at C.prototype (which is a particular instance of B2).
  3. Is "method" an own prop of C.prototype? No.
  4. Ascend the prototype chain, look at B2. [This is a bit weird in OTI: calling `getPrototypeObject` on C.prototype returns B2, when ostensibly it should be B2.prototype, seeing as C.prototype is after all a specific instance of B2. Ends up not making a difference in this case, though.]
  5. Is "method" an own prop of B2? No.
  6. Ascend the prototype chain, look at B2.prototype.
  7. Is "method" an own prop of B2.prototype? No.
  8. Ascend the prototype chain, look at A.
  9. Is "method" an own prop of A? No. (Since we don't look at prototype properties.)
  10. Ascend the prototype chain, look at A.prototype.
  11. Is "method" an own prop of A.prototype? Yes!
  12. Return A.prototype.
And then later on, this A.prototype is normalized "_down_" to A (i.e. the A instance type), and that is what the `ownerType` ends up being in `CheckAccessControls::checkNonOverriddenPropertyVisibility`.

In NTI, however, this is what happens instead:
  1. Is "method" an own prop of C? No. (Since it is defined neither in C nor C.prototype.)
  2. Ascend the prototype chain, look at C.prototype, i.e. a particular instance of B2.
  3. Is "method" an own prop of C.prototype? No. (Since it is defined neither in B2 nor B2.prototype.)
  4. Ascend the prototype chain, look at B2.prototype, i.e. a particular instance of A. [Note that `getPrototypeObject` in NTI is more sane: calling it on C.prototype returns B2.prototype, which makes sense.]
  5. Is "method" an own prop of B2.prototype? Yes, actually. This is because B2.prototype is a special instance of A, and since we look in prototype properties as well, we find "method" as an own prop of A.prototype.
  6. Return B2.prototype.
This then later gets normalized "_up_" (as opposed to "down", in keeping with the caveat) again to A, and so again `ownerType` ends up being A, and all is well. If not for the caveat, i.e. we managed to normalize "down" to B2, we'd get the wrong behavior.

I think it's fair to say that `getOwnProp` is not doing the right thing in NTI, since "own" properties are perfectly well-defined and they do _not_ include properties on the prototype, instead being only the properties defined "directly" on a given object. See e.g. [`Object.prototype.getOwnPropertyNames`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/getOwnPropertyNames) as present in the standard.

So what's the fix here? Well,`getOwnProp` in NTI should be fixed not to look at prototype properties. Of course, the situation isn't quite as simple as that. `getOwnProp` is already defined in `RawNominalType` and called in a bunch of places, and I don't believe it will be too easy to change, since it seems that several callers already implicitly rely on this behavior (i.e. expect `protoProps` to be included). But we should at least change the `getOwnProp` methods in FunctionTypeI to be correct, say by making them eventually call a new method `getOwnPropNoProto` (or whatever) in `RawNominalType`, which would be the same as `getOwnProp` except it doesn't look at `protoProps`. This will fix the behavior of compiler passes that call it, such as CheckAccessControls.

This in itself is easy to do, but we can't do it straightaway -- because of the caveat of `normalizeMaybePrototypeObject`. If we fix only the caveat or only `getOwnProp` then we will get the wrong behavior (as explained in the example). So we have to wait until we can fix both and land both together. And the blocker there (of the two) is the caveat, since that's the one that isn't obvious.

**Comment by aravind-pg** [comment]
FWIW, I actually think the caveat might not be so hard to fix. It should suffice to simply add an `owner` field to `ObjectType` that is null for everyone except prototype objects, for whom it points instead to their owner constructor. This is enough to allow us to implement `getOwnerFunction` right.

**Comment by aravind-pg** [comment]
cc @dimvar @blickly. Sorry for the huge wall of text here, I just wanted to be as clear as I possibly could. Note that I'd be happy to open a PR to land the fixes for this issue (both for the `normalizeMaybePrototypeObject`/`getOwnerFunction` caveat as well as the `getOwnProp` business (hopefully no hidden issues emerge even after those fixes!). I just wanted to ensure we were all on the same page first.

**Comment by dimvar** [comment]
I'm looking at this today, and it seems that for the example in your first comment, Java has the same behavior as NTI. It warns for `b1.method();` in `C`.
